### ISSUE TITLE
trafficlights.getNextSwitch function problem in traci4matlab

### ISSUE THREAD

**Issue by Hosinam** [pipeacosta/traci4matlab#8]
I am working on implementing a control strategy using TRACI4MATLAB in sumo. However I have trouble calculating the remaining phase time for the traffic lights. I use the command "traci.trafficlights.getNextSwitch('0')*0.001-traci.simulation.getTime()", unfortunately it returns the wrong value. I have tracked the problem and I'm guessing that something is wrong with the "traci.trafficlights.getNextSwitch" function. By the way I have defined a switch time program for the traffic light using a WAUT. I will be glad to receive any recommendations to help solving this problem.

**Comment by shuowang572875852** [comment]
Can you give me some help for learning traci4matlab. I want to use  traci.vehicle.subscribeContext  to get  status information about the vehicles on the left and right lanes of the current vehicle., but the return values are error,the following are my codes:
clear
close all
clc

import traci.constants
scenarioPath=['F:\software\SUMO配置文件\bin\quickstart\quickstart.sumocfg'];
system(['sumo-gui -c ' '"' scenarioPath '"' ' --remote-port 8813 --start&']);
 traci.init

SIM_STEPS = [1 900];
beginTime = SIM_STEPS(1);
duration =  SIM_STEPS(2);
endTime =  SIM_STEPS(1) +  SIM_STEPS(2) - 1;
for i = 1:duration
    traci.simulation.step();
  IDList{i,1} = traci.vehicle.getIDList();
%    arrivedIDList= traci.simulation.getArrivedIDList();
  for j=1:length(IDList{i,1})
      [leaderVehicleID{i,1}, dist{i,1}] = traci.vehicle.getLeader(char(IDList{i,1}(1:end,j)),50);
      accel{i,1}(j)=traci.vehicle.getAcceleration(char(IDList{i,1}(1:end,j)));
      speed{i,1}(j)=traci.vehicle.getSpeed(char(IDList{i,1}(1:end,j)));
      laneID{i,1}=traci.vehicle.getLaneID(char(IDList{i,1}(1:end,j)));
      lastStepMeanSpeed{i,1}= traci.lane.getLastStepMeanSpeed(char(laneID{i,1}));
%       length{i,1} = traci.lane.getLength(char(laneID{i,1}));
       allowedSpeed{i,1}(j) = traci.vehicle.getAllowedSpeed(char(IDList{i,1}(1:end,j)));
       laneIndex{i,1} = traci.vehicle.getLaneIndex(char(IDList{i,1}(1:end,j)));
       roadID{i,1} = traci.vehicle.getRoadID(char(IDList{i,1}(1:end,j)));
       lanePosition{i,1}(j) = traci.vehicle.getLanePosition(char(IDList{i,1}(1:end,j)));
        minGap{i,1}(j) = traci.vehicle.getMinGap(char(IDList{i,1}(1:end,j)));
      angle{i,1}(j) = traci.vehicle.getAngle(char(IDList{i,1}(1:end,j)));
      
      % Subscribe to the vehicle ID around the left lane of the current vehicle
      traci.vehicle.subscribeContext(char(IDList{i,1}(1:end,j)),traci.constants.CMD_GET_VEHICLE_VARIABLE,150,{traci.constants.VAR_LANEPOSITION, traci.constants.VAR_SPEED});
       traci.vehicle.addSubscriptionFilterLanes(1);
      traci.vehicle.addSubscriptionFilterLeadFollow(1);
       leftcontextSubscriptionResults{i,1}=traci.vehicle.getContextSubscriptionResults(char(IDList{i,1}(1:end,j))); 
       leftsurround_vehicle{i,1}=keys(leftcontextSubscriptionResults{i,1});
       %  Subscribe to the vehicle ID around the right lane of the current vehicle
       traci.vehicle.subscribeContext(char(IDList{i,1}(1:end,j)),traci.constants.CMD_GET_VEHICLE_VARIABLE,150,{traci.constants.VAR_LANEPOSITION, traci.constants.VAR_SPEED});
        traci.vehicle.addSubscriptionFilterLanes(-1);
        traci.vehicle.addSubscriptionFilterLeadFollow(-1);
       rightcontextSubscriptionResults{i,1}=traci.vehicle.getContextSubscriptionResults(char(IDList{i,1}(1:end,j))); 
       rightsurround_vehicle{i,1}=keys(rightcontextSubscriptionResults{i, 1});
        %  Subscribe to the vehicle ID around the current lane of the current vehicle
       traci.vehicle.subscribeContext(char(IDList{i,1}(1:end,j)),traci.constants.CMD_GET_VEHICLE_VARIABLE,150,{traci.constants.VAR_LANEPOSITION, traci.constants.VAR_SPEED});
        traci.vehicle.addSubscriptionFilterLanes(0);
        traci.vehicle.addSubscriptionFilterLeadFollow(0);
       CurrentcontextSubscriptionResults{i,1}=traci.vehicle.getContextSubscriptionResults(char(IDList{i,1}(1:end,j))); 
       Currentsurround_vehicle{i,1}=keys(CurrentcontextSubscriptionResults{i, 1});
  end
end
traci.close()
But the returned values of Currentsurround_vehicle, rightsurround_vehicle, leftsurround_vehicle are the same,can you give me some help for it. And can you leave a mailbox to discuss it?Looking forward to your reply.

**Comment by bjyurkovich** [comment]
For those that are still having this problem, I was able to trace this down to an `int32` vs `double` conversion problem.

**TL;DR:** This is a serious bug, and it has to do with typecasting.

### More info
It comes down to [this line](https://github.com/pipeacosta/traci4matlab/blob/b89f4827b1239ded3e2ad2abc456c03266292446/%2Btraci/%2Btrafficlights/getUniversal.m#L22), which resolves to the `@readInt` function handle, and is handled as an `int32`.  The problem is, the SUMO specification gives the last bytes of the traffic light protocol (`0x2d`) to be of type `double`.

### Fix
I will try to find the time to do a proper PR, but in the meantime, the quick (not scalable or correct!) fix for this is to create a new function to replace [`+traci/+trafficlights/getUniversal.m`](https://github.com/pipeacosta/traci4matlab/blob/master/%2Btraci/%2Btrafficlights/getUniversal.m) with a new file:

```Matlab
function returnedValue = getUniversalNextLightFix(varID, tlsID)
import traci.constants
global tlsSubscriptionResults

if isempty(tlsSubscriptionResults)
    returnValueFunc = traci.RETURN_VALUE_FUNC.trafficlights;
else
    returnValueFunc = tlsSubscriptionResults.valueFunc;
end

% Prepare the outgoing message and read the response. The result variable
% is a traci.Storage object
result = traci.sendReadOneStringCmd(constants.CMD_GET_TL_VARIABLE,varID,tlsID);
% handleReturValueFunc = str2func(returnValueFunc(varID));

% Use the proper method to read the variable of interest from the result
returnedValue = result.readDouble();  % THIS IS THE FIXED LINE

```

And then call this function (`getUniversalNextLightFix`) in [`getNextSwitch `](https://github.com/pipeacosta/traci4matlab/blob/master/%2Btraci/%2Btrafficlights/getNextSwitch.m), call your updated function:

```Matlab
function nextSwitch = getNextSwitch(tlsID)

import traci.constants
nextSwitch = traci.trafficlights. getUniversalNextLightFix(constants.TL_NEXT_SWITCH, tlsID);
```

**Warning**: It returns the number of seconds (not `ms`), so watch for that.

Hope that helps.

cc @Hosinam (if not too late)

FYI @pipeacosta - I am happy to submit a PR, but if you can suggest where you do the function handle mapping for the functions and best way to test for your standard, that would be helpful to cut down on unit test time.

**Comment by atperin** [comment]
Thank you very much, @bjyurkovich! I've been looking for a fix like this for quite a while!

**Comment by tonmoy-TS** [comment]
Thank you so much. It worked for me. 💯

**Comment by tonmoy-TS** [comment]
I believe, the same problem happens when you call the `traci.trafficlights.getPhaseDuration('tlsID')` function. The same solution applies. 
Just call the updated `getUniversalNextLightFix()` function in place of `getUniversal()`:

```
function phaseDuration = getPhaseDuration(tlsID)

    import traci.constants

    % The following line has been commented out
    % phaseDuration = traci.trafficlights.getUniversal(constants.TL_PHASE_DURATION, tlsID);

    phaseDuration = traci.trafficlights.getUniversalNextLightFix(constants.TL_PHASE_DURATION, tlsID);

end

```

**Comment by bjyurkovich** [comment]
Good catch @tonmoy-TS .  I suspect anywhere the `getUniversal` function is called w.r.t. traffic lights would have this issue (and resulting fix).

I'm a little surprised this has been open for so long.  I have not really dug into it for the last number of years, but it may be because SUMO/TraCI AOI versions are inconsistent, and it would cause a breaking change for an older version... (?? I have no idea... Just a lazy guess).

Edit: ok, I looked into it, and it appears @pipeacosta attempted to address something similar in this commit in 2021: https://github.com/pipeacosta/traci4matlab/commit/a2fb0f19655cc29f05d24ea68d540e605fa19c34

Unfortunately, it appears to fall a bit short in the final reading of the specific traffic light case... I'm going to refrain from a PR because there appears to be something deeper going on that I don't have the time to dig into and do it justice...
